app/docs/src/how-tos/create-templates.md

```diff
@@ -0,0 +1,551 @@
+
+```shellscript
+si template run vpc-pattern.ts --key test -i vpc-pattern-inputs.yaml --dry-run
+```
+
+### Add a Name Pattern
+
+Use `namePattern` to rename components based on input variables:
+
+```typescript
+import { TemplateContext } from "jsr:@systeminit/si";
+import { z } from "npm:zod@4";
+
+export default async function (ctx: TemplateContext) {
+  const inputSchema = z.object({
+    environment: z.string().default("dev"),
+    vpcCidr: z.string().default("10.0.0.0/16"),
+  });
+  type Inputs = z.infer<typeof inputSchema>;
+
+  ctx.inputs(inputSchema);
+
+  ctx.search([
+    'schema:"AWS::EC2::VPC"',
+    'schema:"AWS::EC2::Subnet"',
+    'schema:"AWS::EC2::RouteTable"',
+  ]);
+
+  // Rename components from "demo-*" to "{environment}-*"
+  ctx.namePattern([
+    { pattern: /demo-(.+)/, replacement: "<%= inputs.environment %>-$1" },
+  ]);
+}
+```
+
+The replacement pattern uses [EJS syntax](https://ejs.co/) to reference input variables. We can see if this works by performing a dry-run, once again.
+
+```shellscript
+si template run vpc-pattern.ts --key test --input vpc-pattern-inputs.yaml --dry-run
+```
+
+### Add a Transform
+
+The `transform` function is where you modify component attributes, create new components, or filter the working set:
+
+```typescript
+import { TemplateContext } from "jsr:@systeminit/si";
+import { z } from "npm:zod@4";
+
+export default async function (ctx: TemplateContext) {
+  const inputSchema = z.object({
+    environment: z.string().default("dev"),
+    vpcCidr: z.string().default("10.0.0.0/16"),
+  });
+  type Inputs = z.infer<typeof inputSchema>;
+
+  ctx.inputs(inputSchema);
+
+  ctx.search([
+    'schema:"AWS::EC2::VPC"',
+    'schema:"AWS::EC2::Subnet"',
+    'schema:"AWS::EC2::RouteTable"',
+  ]);
+
+  ctx.namePattern([
+    { pattern: /demo-(.+)/, replacement: "<%= inputs.environment %>-$1" },
+  ]);
+
+  ctx.transform(async (workingSet, inputs) => {
+    const typedInputs = inputs as Inputs;
+
+    // Update VPC CIDR block from inputs
+    const vpc = workingSet.find((c) => c.schemaName === "AWS::EC2::VPC");
+    if (vpc) {
+      await ctx.ensureAttribute(vpc, "/domain/CidrBlock", typedInputs.vpcCidr);
+    }
+
+    // Add environment tags to all components
+    for (const component of workingSet) {
+      await ctx.ensureArrayAttribute(
+        component,
+        "/domain/Tags",
+        (e) => e.subpath === "Key" && e.value === "Environment",
+        { Key: "Environment", Value: typedInputs.environment },
+        { skipIfMissing: true }
+      );
+    }
+
+    return workingSet;
+  });
+}
+```
+
+Perform one final dry-run test.
+
+```shellscript
+si template run vpc-pattern.ts --key test --input vpc-pattern-inputs.yaml --dry-run
+```
+
+### Add External Component References
+
+For templates to work across different workspaces, you need to handle component references (like AWS Credentials and Regions) as inputs rather than hardcoded IDs. Use `SubscriptionInput` for this:
```

Comment:
No need to mention hard coded id's - instead, explain how the subscriptions get resolved. By default, if a subscription is not in the working set, the templating system will attempt to use it unmodified - which won't work, obivously, when you switch workspaces.